apps/rush-lib/src/data/CommonVersionsConfiguration.ts

```diff
@@ -0,0 +1,207 @@
+   */
+  public get preferredVersions(): Map<string, string> {
+    return this._preferredVersions.protectedView;
+  }
+
+  /**
+   * A table of specifies preferred versions maintained by the XStitch tool.
+   *
+   * @remarks
+   * This property has the behavior as the "preferredVersions" property, except these entries
+   * are automatically managed by the XStitch tool.  It is an error for the same dependency name
+   * to appear in both tables.
+   */
+  public get xstitchPreferredVersions(): Map<string, string> {
+    return this._xstitchPreferredVersions.protectedView;
+  }
+
+  /**
+   * For a given dependency, a list of SemVer ranges that will be ignored by "rush check".
+   *
+   * @remarks
+   * The "rush check" command can be used to enforce that every project in the repo
+   * must specify the same SemVer range for a given dependency.  However, sometimes
+   * exceptions are needed.  The allowedAlternativeVersions table specifies a list of
+   * other SemVer ranges that will be ignored by "rush check" for a given dependency.
+   * Note that the normal version range (used by the majority of projects) should NOT be
+   * included in this list.
+   */
+  public get allowedAlternativeVersions(): Map<string, ReadonlyArray<string>> {
+    return this._allowedAlternativeVersions.protectedView;
+  }
+
+  private constructor(commonVersionsJson: ICommonVersionsJson | undefined, _filename: string) {
+    this._preferredVersions = new ProtectableMap<string, string>(
+      { onSet: this._onSetPreferredVersions.bind(this) });
+
+    this._xstitchPreferredVersions = new ProtectableMap<string, string>(
+      { onSet: this._onSetPreferredVersions.bind(this) });
+
+    this._allowedAlternativeVersions = new ProtectableMap<string, string[]>(
+      { onSet: this._onSetAllowedAlternativeVersions.bind(this) });
+
+    if (commonVersionsJson) {
+      CommonVersionsConfiguration._deserializeTable(this.preferredVersions,
+        commonVersionsJson.preferredVersions);
+      CommonVersionsConfiguration._deserializeTable(this.xstitchPreferredVersions,
+        commonVersionsJson.xstitchPreferredVersions);
+      CommonVersionsConfiguration._deserializeTable(this.allowedAlternativeVersions,
+        commonVersionsJson.allowedAlternativeVersions);
+    }
+
+    this._filename = _filename;
+  }
+
+  private _onSetPreferredVersions(source: ProtectableMap<string, string>, key: string, value: string): string {
+    PackageName.validate(key);
+
+    if (source === this._preferredVersions) {
+      if (this._xstitchPreferredVersions.has(key)) {
+        throw new Error(`The package "${key}" cannot be added to preferredVersions because it was already`
+          + ` added to xstitchPreferredVersions`);
+      }
+    } else {
+      if (this._preferredVersions.has(key)) {
+        throw new Error(`The package "${key}" cannot be added to xstitchPreferredVersions because it was already`
+          + ` added to preferredVersions`);
```

Comment:
Do you want to throw for each collision, or collect them all and throw at the end?

Reply:
This is a good idea, but it would necessarily separate "loading the file" from "validating the file contents" which is a little complicated.  I'm over time already for this PR, so let's procrastinate it until we have evidence that anybody is annoyed by the simplistic approach.

---
In reply to: [178636560](https://github.com/Microsoft/web-build-tools/pull/593#discussion_r178636560) [](ancestors = 178636560)